src/Binance/Entry.cpp

```diff
@@ -29,27 +30,28 @@ string Entry::signJSON(TWCoinType coin, const std::string& json, const Data& key
     return Signer::signJSON(json, key);
 }
 
-HashPubkeyList Entry::preImageHashes(TWCoinType coin, const Data& txInputData) const {
-    auto input = Proto::SigningInput();
-    Data preImageHash;
-    if (input.ParseFromArray(txInputData.data(), (int)txInputData.size())) {
-        const auto signer = Signer(input);
-        // return preimage hash and dummy pubkeyhash (only one signature anyways)
-        preImageHash = signer.preImageHash();
-    }
-    return HashPubkeyList{std::make_pair(preImageHash, Data())};
+Data Entry::preImageHashes(TWCoinType coin, const Data& txInputData) const {
+    return txCompilerTemplate<Proto::SigningInput, TxCompiler::Proto::PreSigningOutput>(
+        txInputData, [](const auto& input, auto& output) {
+            Signer signer(input);
+
+            auto preImageHash = signer.preImageHash();
+            auto preImage = signer.signaturePreimage();
+            output.set_datahash(preImageHash.data(), preImageHash.size());
+            output.set_data(preImage.data(), preImage.size());
+        });
 }
 
 void Entry::compile(TWCoinType coin, const Data& txInputData, const std::vector<Data>& signatures, const std::vector<PublicKey>& publicKeys, Data& dataOut) const {
-    auto input = Proto::SigningInput();
-    if (input.ParseFromArray(txInputData.data(), (int)txInputData.size())) {
-        const auto signer = Signer(input);
-        if (signatures.size() == 1 && publicKeys.size() == 1) {
-            const auto output = signer.compile(signatures[0], publicKeys[0]);
-            const auto serializedOut = output.SerializeAsString();
-            dataOut.insert(dataOut.end(), serializedOut.begin(), serializedOut.end());
-        }
-    }
+    dataOut = txCompilerTemplate<Proto::SigningInput, Proto::SigningOutput>(
+        txInputData, [&](const auto& input, auto& output) {
+            if (signatures.size() != 1 || publicKeys.size() != 1) {
+                output.set_errorcode(Common::Proto::Error_no_support_n2n);
```

Comment:
Empty array(s) also will result in no_support_n2n, which is a bit misleading

Reply:
Yes, will check empty